### PR TITLE
Support for POST requests in C# and Javascript

### PR DESCRIPTION
**Pull request by andrueastman**
This PR adds support for snippet generation for POST requests sent to the API for C# and JavaScript(#10 and #22)

C# code generation involves the use of an object generation function that is language specific while JavaScript involves minor modification of the JSON string in the request to create the object string. 

In addition, the following enhancements/fixes/improvements are made
- Added Unit tests to improve coverage :smiley:
- Fixed quotation marks present in Skip and Top language expressions for JavaScript and C#
- Use of the "CommonGenerator.GetListAsStringForSnippet" function where possible to reduce instances of repeated code.

### CODE REVIEW

File: GraphWebApi/Controllers/GraphExplorerSnippetsController.cs [hunk 1]
```diff
@@ -41,6 +42,7 @@ public IActionResult Get(string arg)
         [Consumes("application/http")]
         public async Task<IActionResult> PostAsync(string lang = "c#")
         {
+            Request.EnableRewind();
             var streamContent = new StreamContent(Request.Body);
```
Comment (darrelmiller) [inline GraphWebApi/Controllers/GraphExplorerSnippetsController.cs]: I'm not sure what EnableRewind does under the covers but if you call `await streamContent.LoadIntoBufferAsync()` it will also make the streamContent object rewindable.

File: CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs [hunk 2]
```diff
@@ -13,47 +16,90 @@ public static class JavaScriptGenerator
...
+                        //eg microsoft.graph.data => data
+                        objectType = objectType.Split(".").Last();
+                        //parameter for the post.
+                        name = $"{{{objectType} : { name }}}";
+                    }
+                             
+                    snippetBuilder.Append($"client.api('{snippetModel.Path}')");
+                    //append beta
+                    snippetBuilder.Append(BetaSectionString(snippetModel.ApiVersion));
+                    snippetBuilder.Append($".post({name});");
```
[the remark this answers is not in the document]
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs]: You raise a good point there. Currently, simple snippets of the following kind is generated. 
(mainly motivated from samples [here](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/dev/samples/browser/src/request.js) )
`
var events = client.api('/me/events/AAMkAGIAAAoZDOFAAA=').get();
`
What would be the best way to do this in your opinion?

**Comment by andrueastman** [comment]
@muthurathinam 
The project is meant to return a snippet when an example request is identified by a tool i.e. apidoctor in the MSGraph docs so that the docs can be updated with example snippets.

You can view an example of the docs with snippets injected [here](https://review.docs.microsoft.com/en-us/graph/api/group-get?view=graph-rest-1.0&branch=andrueastman%2Fsnippets-preview-get-post-js-csharp&tabs=Javascript).

I have also made a tiny update with markdown files to show the example snippets currently generated. Any more feedback on them is much appreciated.

**Comment by muthurathinam** [comment]
@andrueastman  
Thank you for providing the details.
#### Here are something that I need clarification with:
* In the example docs page that you have highligted, it is mentioned as JavaScript snippet but we are providing TypeScript snippet. Is it something doesn't align ?
* In the same docs, how would the user know the snippet should be used with the SDK ?

#### Here are my suggestions:
* In the [generated snippet](https://github.com/microsoftgraph/microsoft-graph-explorer-api/blob/andrueastman/dev-refactor/javascript-examples.md#snippet-generated), I see `var`, `let` and `const` are all used. so our intent is to provide shining latest JavaScript snippet its better we drop `var` and use `let` and `const` accordingly.
* In the same generated snippet `const` should be used instead of `let`. (only the variables that are going to change in future code should be declared with `let`)
* `client.api().get/post/anyAction()` returns promise so the code that we generated will not be appropriate. refer the example snippets in the SDK repo to get more clarity on this. [#1](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/dev/docs/Actions.md), [#2](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/dev/docs/OtherAPIs.md), [#3](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/dev/docs/QueryParameters.md), [#4](https://github.com/microsoftgraph/msgraph-sdk-javascript/tree/dev/docs)

File: CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (muthurathinam) [reply CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs]: Refer https://github.com/microsoftgraph/microsoft-graph-explorer-api/pull/24#issuecomment-478840379

**Comment by andrueastman** [comment]
@muthurathinam 
Thanks for the pointers on the snippets.

The docs will be updated with other hints and links to a show how to do other stuff like auth. 

I have made an updated taking in your feedback with regards to the JavaScript variables and futures. You can check out the docs preview of the snippets [here](https://review.docs.microsoft.com/en-us/graph/api/user-post-users?branch=andrueastman%2Fsnippets-preview-get-post-js-csharp&view=graph-rest-1.0&tabs=Javascript) and/or view the updated [generated snippet](https://github.com/microsoftgraph/microsoft-graph-explorer-api/blob/andrueastman/dev-refactor/javascript-examples.md#snippet-generated).

Looking forward to your thoughts.

**Comment by andrueastman** [comment]
@muthurathinam 

Well spotted. This should now be resolved if you follow the same link.

File: c-sharp-examples.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+Host: graph.microsoft.com
+Prefer: outlook.timezone="Pacific Standard Time"
+```
+
+#### Example Get Request snippet generated
+
+```cs
+GraphServiceClient graphClient = new GraphServiceClient( authProvider );
+
+var events = await graphClient.Me.Events["AAMkAGIAAAoZDOFAAA="]
```
Comment (pwombwa) [inline c-sharp-examples.md]: Can we ensure that the variable name reflects the expected output? In this example, the generated variable name should be `event` as opposed to `events` since we are getting a specific event.

File: CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs [hunk 4]
```diff
@@ -13,59 +16,117 @@ public static class JavaScriptGenerator
...
+        {
+            return apiVersion.Equals("beta") ? "\n\t\t.version('beta')" : "";
+        }
+
+        /// <summary>
+        /// Generate the snippet section that makes the call to the api with the surrounding try catch block
+        /// </summary>
+        /// <param name="snippetModel">Snippet model built from the request</param>
+        /// <param name="actions">String of actions to be done inside the code block</param>
+        private static string GenerateRequestSection(SnippetModel snippetModel, string actions)
```
Comment (MIchaelMainer) [inline CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs]: Applies to all of GenerateRequestSection()

* don't apply code comments for code snippets. The snippet will have plenty of context from the surrounding documentation. Code comments are more useful for larger code blocks 
* try-catch-throw does not provide any value for a snippet unless the snippet demonstrates an expected error.
* console.log does not provide much value.  

Code snippets provide the following benefits:
* indexable strings that customers can use to find information, which also supports,
* discovery of the existence of the client libraries, and the associated feature documentation.
* copy and pasteable snippets to use in projects.

File: c-sharp-examples.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
+---
+title: "C Sharp examples"
+description: "Examples Snippets generated by the api"
+author: "andrueastman"
+---
+
```
Comment (MIchaelMainer) [inline c-sharp-examples.md]: These examples bring me joy

File: CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs [hunk 6]
```diff
@@ -83,19 +86,116 @@ public static string GenerateQuerySection(SnippetModel snippetModel, LanguageExp
...
+        /// <summary>
+        /// Function to find the name of the type/class being used by identifier that is the last item in the path collection. This function gets
+        /// the right <see cref="IEdmType"/> that is is to be used for the search based on the segment type>
+        /// </summary>
+        /// <param name="oDataPathSegment">Odata path segment that is the root of the search </param>
+        /// <param name="path">List of string that show the depth of the search into the definition from the odataPath type definition</param>
+        public static string GetClassNameFromIdentifier(ODataPathSegment oDataPathSegment, ICollection<string> path )
+        {
+            switch (oDataPathSegment)
+            {
```
[the remark this answers is not in the document]
Comment (MIchaelMainer) [reply CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs]: I'm okay with adding these to the backlog.

File: CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs [hunk 7]
```diff
@@ -83,19 +86,116 @@ public static string GenerateQuerySection(SnippetModel snippetModel, LanguageExp
...
+                return elementDefinition.ToString();
+            }
+
+            //the second element in the path is the property we are searching for
+            var searchIdentifier = searchPath.ElementAt(1);
+            
+            //Loop through the properties of the entity if is structured
+            if (elementDefinition is IEdmStructuredType structuredType)
+            {
+                foreach (var property in structuredType.DeclaredProperties)
```
[the remark this answers is not in the document]
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs]: Well spotted. 

This should be resolved with the latest changes.

File: CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs (hunk 4 above)
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs]: Resolved in latest updates.

File: c-sharp-examples.md (hunk 3 above)
Comment (andrueastman) [reply c-sharp-examples.md]: This is now resolved with the latest updates.

File: CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs]: This is resolved with the latest changes

File: CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MIchaelMainer) [reply CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs]: Can you provide me a link to the resolution?

File: CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs [hunk 8]
```diff
@@ -13,59 +16,117 @@ public static class JavaScriptGenerator
...
+                else if (snippetModel.Method == HttpMethod.Post)
+                {
+                    var name = "";
+                    // create variable to send out if we have a body
+                    if (!string.IsNullOrEmpty(snippetModel.RequestBody))
                     {
-                        snippetBuilder.Append("\n\t.version('beta')");
+                        name = snippetModel.ResponseVariableName;
+                        snippetBuilder.Append(JavascriptGenerateObjectFromJson(snippetModel.RequestBody,name));
+                        var objectType = CommonGenerator.GetClassNameFromIdentifier(snippetModel.Segments.Last() , new List<string> { snippetModel.Segments.Last().Identifier });
```
[the remark this answers is not in the document]
Comment (MIchaelMainer) [reply CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs]: This is still a todo, I think.

File: CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs]: @MIchaelMainer You can check it out [here](https://github.com/microsoftgraph/microsoft-graph-explorer-api/blob/1d9281137113f7a53880dbbff004a78396b9949d/CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs#L176).
Also added a test case to cover the edge case [here](https://github.com/microsoftgraph/microsoft-graph-explorer-api/blob/1d9281137113f7a53880dbbff004a78396b9949d/CodeSnippetsReflection.Test/CommonGeneratorShould.cs#L288)

File: CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs]: @MIchaelMainer As discussed, I think we can resolve this and raise an issue for the snippet generation for the C Sharp structural properties as the JavaScript structural structural properties can be generated fine this way.

[the remark this answers is not in the document]
Comment (andrueastman) [reply CodeSnippetsReflection/LanguageGenerators/JavaScriptGenerator.cs]: Issue #25 will be used to resolve this

File: CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MIchaelMainer) [reply CodeSnippetsReflection/LanguageGenerators/CommonGenerator.cs]: I wouldn't call this an edge case. All entity types inherit from at least one base entity.

**Comment by darrelmiller** [comment]
There are some issues in this sample:

```csharp
GraphServiceClient graphClient = new GraphServiceClient( authProvider );

var value = "NV001";

var name = "x-custom-header-group-id";

var valueVar = "Nevada";

var nameVar = "x-custom-header-group-nameVar";

//create InternetMessageHeader list and populate it
var internetMessageHeaders = new List<InternetMessageHeader>();
internetMessageHeaders.Add(new InternetMessageHeader(nameVar,valueVar));
internetMessageHeaders.Add(new InternetMessageHeader(name,value));

//create instance of EmailAddress
var emailAddress = new EmailAddress
{
	Address = "AlexW@contoso.OnMicrosoft.com",
};

//create Recipient list and populate it
var toRecipients = new List<Recipient>();
toRecipients.Add(new Recipient(emailAddress));

//create instance of ItemBody
var body = new ItemBody
{
	ContentType = "HTML",
	Content = "The group represents Nevada.",
};

//create instance of Message
var message = new Message
{
	Subject = "9/9/2018: concert",
	Body = body,
	ToRecipients = toRecipients,
	InternetMessageHeaders = internetMessageHeaders,
};

await graphClient.Me
	.SendMail(message,saveToSentItems)
	.Request()
	.PostAsync()
```

- Recipient does not have a ctor that accepts email address.  You need to set the property explictly
- ContentType is an enum not a string.
